### ISSUE TITLE
`Cow<[T]>` layout forces unnecessary branching

### ISSUE THREAD

**Issue by Jules-Bertholet** [rust-lang/rust#117763]
I tried this code ([Godbolt](https://rust.godbolt.org/z/GfhnT65ne)):

```rust
pub fn test<'a>(cow: &'a std::borrow::Cow<'a, [u8]>) -> &'a [u8] {
    &*cow
}
```

I expected to see this happen: The generated assembly is branchless. The pointer and length are at the same offset in both variants of the `Cow`.

Instead, this happened: There is a branch in the generated assembly. The pointer to the slice is at a different offset in the `Owned` and `Borrowed` variants of the `Cow`.

---

The ideal layout for `Cow<[T]>` (and `Cow<str>`) would look like this:

```
If `T` is not a ZST:
┌──────────┬──────────┬────────────────────────────────────────────────────┐
│ pointer  │ length   │ capacity if `Owned`, or `usize::MAX` if `Borrowed` │
└──────────┴──────────┴────────────────────────────────────────────────────┘

If `T` is a ZST:
┌──────────┬──────────┬────────────────────────────────────────┐
│ pointer  │ length   │ boolean flag for `Owned` vs `Borrowed` │
└──────────┴──────────┴────────────────────────────────────────┘
```

The present layout looks like this:

```
`Owned` variant:
┌──────────┬──────────┬──────────┐
│ pointer  │ capacity │ length   │
└──────────┴──────────┴──────────┘

`Borrowed` variant:
┌──────────┬──────────┬──────────┐
│ 0x0      │ pointer  │ length   │
└──────────┴──────────┴──────────┘
```

Because of this non-optimal layout, every access to the contained pointer requires a branch on the enum variant.

(Fixing #45431 is a prerequisite to fixing this)

### Meta

`rustc --version --verbose`:
```
rustc 1.75.0-nightly (189d6c71f 2023-11-06)
binary: rustc
commit-hash: 189d6c71f3bb6c52113b5639a80839791974fd22
commit-date: 2023-11-06
host: x86_64-unknown-linux-gnu
release: 1.75.0-nightly
LLVM version: 17.0.4
```

@rustbot label A-layout I-heavy I-slow T-compiler T-libs

**Comment by the8472** [comment]
It wouldn't solve #45431, but #106790 would add a larger niche to RawVec which could be used for the flag. Though that might actually turn it into `((cap, ptr), len)`... which should also work for this purpose because then the flag can be stored in the first field before the borrowed payload.

**Comment by DaniPopes** [comment]
Looks like this is fixed with #106790 in the latest nightly (Godbolt link in OP):

```gas
example::test:
        mov     rax, qword ptr [rdi + 8]
        mov     rdx, qword ptr [rdi + 16]
        ret
```

```
print-type-size type: `alloc::borrow::Cow<'_, [u8]>`: 24 bytes, alignment: 8 bytes
print-type-size     variant `Borrowed`: 24 bytes
print-type-size         padding: 8 bytes
print-type-size         field `.0`: 16 bytes, alignment: 8 bytes
print-type-size     variant `Owned`: 24 bytes
print-type-size         field `.0`: 24 bytes
```

**Comment by RalfJung** [comment]
Is there a test anywhere checking for this? Subtle codegen/layout issues of this sort have a high chance of regressing, so we shouldn't close it without a test.

**Comment by the8472** [comment]
only partially, https://github.com/rust-lang/rust/blob/733dab558992d902d6d17576de1da768094e2cf3/tests/ui/structs-enums/type-sizes.rs#L338-L340 checks the vec layout, but not that the cow -> slice transform is branchless.

**Comment by RalfJung** [comment]
Yeah, having that layout doesn't enforce that the fields are truly at the same offset. And TBH I could imagine entirely legitimate changes to the layout algorithm that would change this in the future.  A test would ensure we know when this happens so we can then decide on the trade-off.
